examples/chat/module/rust/src/lib.rs

```diff
@@ -14,137 +14,76 @@
+        Ok(())
+    })
+    .unwrap_or(Err(oak::OakStatus::ERR_INTERNAL))
+    .err()
+    .unwrap_or(oak::OakStatus::OK)
+    .value()
 }
 
-fn room_id_duplicate_err<T>() -> grpc::Result<T> {
-    Err(grpc::build_status(
-        grpc::Code::ALREADY_EXISTS,
-        "room already exists",
-    ))
-}
-
-impl ChatNode for Node {
-    fn create_room(&mut self, req: CreateRoomRequest) -> grpc::Result<Empty> {
-        info!("creating room");
-        if self.rooms.contains_key(&req.room_id) {
-            return room_id_duplicate_err();
-        }
-
-        // Create a new Node for this room, and keep the write handle and admin token in the `rooms`
-        // map.
-        self.rooms.insert(req.room_id, Room::new(req.admin_token));
-
-        Ok(Empty::new())
-    }
-
-    fn destroy_room(&mut self, req: DestroyRoomRequest) -> grpc::Result<Empty> {
-        info!("destroying room");
-        match self.rooms.entry(req.room_id) {
-            Entry::Occupied(e) => {
-                if e.get().admin_token == req.admin_token {
-                    // Close the only input channel that reaches the per-room Node, which
-                    // will trigger it to terminate.
-                    e.get().sender.close().expect("could not close channel");
-                    e.remove();
-                    Ok(Empty::new())
-                } else {
-                    Err(grpc::build_status(
-                        grpc::Code::PERMISSION_DENIED,
-                        "invalid admin token",
-                    ))
-                }
-            }
-            Entry::Vacant(_) => room_id_not_found_err(),
+impl Router {
+    fn event_loop(mut self, in_handle: u64) -> Result<(), oak::OakError> {
+        let in_channel = oak::ReadHandle {
+            handle: oak::Handle::from_raw(in_handle),
+        };
+        let receiver = oak::io::Receiver::new(in_channel);
+        info!("starting event loop");
+        loop {
+            let grpc_invocation: Invocation = receiver.receive()?;
+            self.handle_grpc_invocation(grpc_invocation)?;
         }
     }
 
-    fn subscribe(&mut self, req: SubscribeRequest, mut writer: grpc::ChannelResponseWriter) {
-        info!("subscribing to room");
-        match self.rooms.get(&req.room_id) {
-            None => {
-                writer
-                    .close(room_id_not_found_err())
-                    .expect("could not close channel");
-            }
-            Some(room) => {
-                info!("new subscription to room {:?}", req.room_id);
-                let command = Command::Join(writer.handle());
-                room.sender
-                    .send(&command)
-                    .expect("could not send command to room Node");
-            }
-        };
+    fn handle_grpc_invocation(&mut self, grpc_invocation: Invocation) -> Result<(), oak::OakError> {
+        let grpc_invocation_label = grpc_invocation.request_receiver.label()?;
```

Comment:
The label is now considered public, since the node is not getting tainted by reading the label.

This means in principle that anyone might get hold of the label. I know this is just a toy example, but at the very least the validity of the label should be checked somewhere (eg that the label can be generated from the bearer token). The label is used for routing to a room, which would mean anyone who has the public label could send messages to the room (or subscribe to it) without knowing the bearer token. A malicious client could manually override the gRPC metadata to use the label.